src/java/org/apache/cassandra/index/sai/iterators/HnswIntersectionIterator.java

```diff
@@ -0,0 +1,241 @@
+ * and because it's really more of an order by than a where clause; HNSW doesn't
+ * "eliminate" any keys, they will all be in the results if you iterate far down enough.
+ *
+ * HnswIntersectionIterator performs what the Pinterest guys call Post Filtering:
+ * we perform the HNSW search, then go through the keys from the other predicates
+ * looking for matches.  If we don't get enough matches, we expand the HNSW search
+ * window and perform the search over again.
+ *
+ * To attempt to minimize the passes, we double the HNSW window with each iteration.
+ */
+public class HnswIntersectionIterator extends KeyRangeIterator
+{
+    private final Builder builder;
+    private final KeyRangeIterator hnswIterator;
+    private HnswOnePassIterator onePassIterator;
+    private final Set<PrimaryKey> hnswMatches = new HashSet<>();
+    private final Set<PrimaryKey> seenMatchingKeys = new HashSet<>();
+    private int hnswBatchSize;
+
+    protected HnswIntersectionIterator(Builder builder)
+    {
+        super(builder.statistics);
+        this.builder = builder;
+        this.hnswIterator = builder.hnswIterator;
+        hnswBatchSize = limit();
+        this.onePassIterator = newOnePassIterator(builder);
+        loadMoreHnswMatches();
+    }
+
+    private HnswOnePassIterator newOnePassIterator(Builder builder)
+    {
+        return new HnswOnePassIterator(builder.limit,
+                                       builder.otherSuppliers.stream().map(Supplier::get).collect(Collectors.toList()));
+    }
+
+    @Override
+    public PrimaryKey skipTo(PrimaryKey nextKey)
+    {
+        return onePassIterator.skipTo(nextKey);
+    }
+
+    @Override
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        // external callers will call skipTo()
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        onePassIterator.close();
+    }
+
+    // TODO do we need to compute all the rows, and then re-sort by PK?
+    @Override
+    protected PrimaryKey computeNext()
+    {
+        // TODO integrate this better with the actual HNSW batch window
+        while (seenMatchingKeys.size() < limit())
+        {
+            while (onePassIterator.hasNext())
+            {
+                var next = onePassIterator.next();
+                if (seenMatchingKeys.add(next))
+                {
+                    return next;
+                }
+            }
+
+            if (!loadMoreHnswMatches()) {
+                return endOfData();
+            } else {
+                hnswBatchSize *= 2;
+            }
+            onePassIterator = newOnePassIterator(builder);
+        }
+        return endOfData();
+    }
+
+    private boolean loadMoreHnswMatches()
+    {
+        for (int i = 0; i < hnswBatchSize; i++)
+        {
+            if (!hnswIterator.hasNext())
+            {
+                return i > 0;
+            }
+            hnswMatches.add(hnswIterator.next());
```

Comment:
right, this is why we need a real planner -- obviously there are scenarios where it would be better to just materialize the non-ANN result and invert the approach.  Can we identify that ahead of time today?